examples/cfd/stokes_mgn/pi_inference.py

```diff
@@ -0,0 +1,525 @@
+            func_arch=False,
+        ).to(
+            self.device
+        )  # For pure inference (no gradients)
+        self.graph_full = Graph(
+            self.nodes,
+            [Key("x"), Key("y")],
+            [
+                Key("u"),
+                Key("v"),
+                Key("p"),
+                Key("continuity"),
+                Key("momentum_x"),
+                Key("momentum_y"),
+            ],
+            func_arch=False,
+        ).to(self.device)
+
+        self.optimizer = torch.optim.Adam(self.model.parameters(), lr=0.001)
+
+    def parabolic_inflow(self, y, U_max=0.3):
+        u = 4 * U_max * y * (0.4 - y) / (0.4**2)
+        v = torch.zeros_like(y)
+        return u, v
+
+    def loss(self):
+        x_in, y_in = self.coords_inflow[:, 0:1], self.coords_inflow[:, 1:2]
+        results_inflow = self.graph_full.forward({"x": x_in, "y": y_in})
+        pred_u_in, pred_v_in = results_inflow["u"], results_inflow["v"]
+
+        x_no_slip, y_no_slip = self.coords_noslip[:, 0:1], self.coords_noslip[:, 1:2]
+        results_noslip = self.graph_full.forward({"x": x_no_slip, "y": y_no_slip})
+        pred_u_noslip, pred_v_noslip = results_noslip["u"], results_noslip["v"]
+
+        x_int, y_int = self.coords[:, 0:1], self.coords[:, 1:2]
+        results_int = self.graph_full.forward({"x": x_int, "y": y_int})
+        pred_mom_u, pred_mom_v, pred_cont = (
+            results_int["momentum_x"],
+            results_int["momentum_y"],
+            results_int["continuity"],
+        )
+        pred_u, pred_v, pred_p = results_int["u"], results_int["v"], results_int["p"]
+
+        u_in, v_in = self.parabolic_inflow(self.coords_inflow[:, 1:2])
+
+        # Compute losses
+        # data loss
+        loss_u = torch.mean((self.gnn_u - pred_u) ** 2)
+        loss_v = torch.mean((self.gnn_v - pred_v) ** 2)
+        loss_p = torch.mean((self.gnn_p - pred_p) ** 2)
+
+        # inflow boundary condition loss
+        loss_u_in = torch.mean((u_in - pred_u_in) ** 2)
+        loss_v_in = torch.mean((v_in - pred_v_in) ** 2)
+
+        # noslip boundary condition loss
+        loss_u_noslip = torch.mean(pred_u_noslip**2)
+        loss_v_noslip = torch.mean(pred_v_noslip**2)
+
+        # pde loss
+        loss_mom_u = torch.mean(pred_mom_u**2)
+        loss_mom_v = torch.mean(pred_mom_v**2)
+        loss_cont = torch.mean(pred_cont**2)
+
+        return (
+            loss_u,
+            loss_v,
+            loss_p,
+            loss_u_in,
+            loss_v_in,
+            loss_u_noslip,
+            loss_v_noslip,
+            loss_mom_u,
+            loss_mom_v,
+            loss_cont,
+        )
+
+    def train(self):
+        (
+            loss_u,
+            loss_v,
+            loss_p,
+            loss_u_in,
+            loss_v_in,
+            loss_u_noslip,
+            loss_v_noslip,
+            loss_mom_u,
+            loss_mom_v,
+            loss_cont,
+        ) = self.loss()
+        loss = (
```

Comment:
Add a comment here for what this weighting is doing

Reply:
Thanks, added.